### PR TITLE
LIBDRS-9044 - Modify namespace values; add unit test.

### PR DESCRIPTION
**Pull request by daveneiman**
**Resolves Issue**: [LIBDRS-9044](https://jira.huit.harvard.edu/browse/LIBDRS-9044)

# What does this Pull Request do?
Fixes two FitsOutput methods that threw an exception due to the upgrade to Jdom2.

# How should this be tested?
There is a new test class.
`mvn clean test -Dtest=FitsOutputTest`
**Cannot build and test via Docker as documented in the README due to many other unrelated test failures. **

A description of what steps someone could take to:
The error being fixed can be seen by rolling back the changes in FitsOutput and running the test class.

# Test coverage
Yes/No: Are changes in this pull-request covered by:
- unit tests? yes
- integration tests? no

# Interested parties
@cvicary @awoods @pwinckles

### CODE REVIEW

File: src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+public class FitsOutputTest extends AbstractLoggingTest {
+
+    @Test
+    public void testFitsOutput() throws Exception {
+
+        FitsOutput fitsOutput = new FitsOutput(fitsMetadata);
+        FitsMetadataElement element = fitsOutput.getMetadataElement("fileinfo");
```
Comment (pwinckles) [inline src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:33]: The output from line 33 is immediately overwritten by the output of line 34. Is this line needed?

File: src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+import edu.harvard.hul.ois.fits.FitsMetadataElement;
+import edu.harvard.hul.ois.fits.FitsOutput;
+import edu.harvard.hul.ois.fits.tests.AbstractLoggingTest;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class FitsOutputTest extends AbstractLoggingTest {
+
+    @Test
+    public void testFitsOutput() throws Exception {
```
Comment (pwinckles) [inline src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: This test appears to pass with or without the changes you made to FitsOutput.

**Comment by pwinckles** [comment]
@daveneiman can you be more specific about what you mean by:

> Cannot build and test via Docker as documented in the README due to many other unrelated test failures

File: src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java (hunk 1 above)
Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java]: This line should have been removed. It's now gone.

File: src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java (hunk 2 above)
Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: Not for me.

Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: Following current Docker build & running tests:
```
# The build only needs to be run once
docker build -f docker/Dockerfile-test -t fits-test .
docker run --rm -v `pwd`:/fits -v ~/.m2:/root/.m2 fits-test mvn clean test
```
Some errors on running tests which appear unrelated to my changes:
```
Running edu.harvard.hul.ois.fits.consolidation.OISConsolidatorTest
[TEST] 2022-12-15 09:03:00 - ERROR - ToolBelt:144 - Thread [1] Error instantiating class: edu.harvard.hul.ois.fits.tools.jhove.Jhove -- Exception thrown: java.lang.reflect.InvocationTargetException -- Error message: null
[TEST] 2022-12-15 09:03:18 - ERROR - ToolBelt:144 - Thread [1] Error instantiating class: edu.harvard.hul.ois.fits.tools.droid.Droid -- Exception thrown: java.lang.reflect.InvocationTargetException -- Error message: null
[TEST] 2022-12-15 09:03:18 - ERROR - Fits:683 - Tool error processing file: image-vectorgraphic.svg
null
java.lang.reflect.InvocationTargetException: null
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[?:?]
	at edu.harvard.hul.ois.fits.tools.ToolBelt.createToolClassInstance(ToolBelt.java:174) ~[classes/:?]
	at edu.harvard.hul.ois.fits.tools.ToolBelt.init(ToolBelt.java:127) ~[classes/:?]
	at edu.harvard.hul.ois.fits.tools.ToolBelt.<init>(ToolBelt.java:56) ~[classes/:?]
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:230) ~[classes/:?]
	at edu.harvard.hul.ois.fits.consolidation.OISConsolidatorTest$3.<init>(OISConsolidatorTest.java:153) ~[test-classes/:?]
	at edu.harvard.hul.ois.fits.consolidation.OISConsolidatorTest.jhoveBeforeDroid(OISConsolidatorTest.java:153) ~[test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
(elided)
Caused by: java.lang.ClassCastException: class org.apache.commons.configuration.SubnodeConfiguration cannot be cast to class org.apache.commons.configuration.HierarchicalConfiguration (org.apache.commons.configuration.SubnodeConfiguration is in unnamed module of loader 'app'; org.apache.commons.configuration.HierarchicalConfiguration is in unnamed module of loader edu.harvard.hul.ois.fits.tools.utils.ParentLastClassLoader$ChildClassLoader @2c5529ab)
	at edu.harvard.hul.ois.fits.tools.utils.XsltTransformMap.getMap(XsltTransformMap.java:35) ~[classes/:?]
	at edu.harvard.hul.ois.fits.tools.jhove.Jhove.<init>(Jhove.java:77) ~[classes/:?]
	... 43 more
```

Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: The reason for this jdom2 fix is that a similar test in drs2_services was also failing.

Comment (pwinckles) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: @daveneiman Interesting. That appears to be a class loader issue. Can you please try changing line 60 in `ParentLastClassLoader` to the following:

```java
loadByParentClassLoader.add("org.apache.commons.configuration");
```

Comment (pwinckles) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: > The reason for this jdom2 fix is that a similar test in drs2_services was also failing.

I don't doubt that there may be an issue here, I would just like to have a reproducible failing test for it.

Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: @pwinckles I modified `ParentLastClassLoader` as request but same errors.

Comment (pwinckles) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: Weird. Can you try turning on the trace logging for that class? It's unclear to me how `org.apache.commons.configuration.HierarchicalConfiguration` could be being loaded on the child classloader after making this change. It's also weird that your system is behaving so unusually, especially considering the tests are being executed from a docker container. Can you try completely rebuilding your test image?

```
docker build --no-cache -f docker/Dockerfile-test -t fits-test .
```

Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: I've run the tests (up until the first couple of errors) with 'trace' on for `ParentLastClassLoader`.
I have a 2GB file for you. How best to send it your way?

Comment (pwinckles) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: If you zip it, how big is it?

Comment (daveneiman) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: I meant 2.2MB (not GB). Zips down to 42K.

Comment (pwinckles) [reply src/test/java/edu/harvard/hul/ois/fits/junit/FitsOutputTest.java:30]: Ah! That's a big difference! :)

If you aren't comfortable attaching inline here, you could send it to me on slack.

**Review by pwinckles — COMMENTED** [review]
Okay, I was able to reproduce the problem now. Not sure what the problem was the other day.

I would recommend the following changes:

1. Pull the namespace off `fitsXml` when it's set and store it at the class level, this is the "un-prefixed" namespace.
2. Create a new "prefixed" namespace based on the namespace extracted from the document
3. Remove all other namespace creations in the file (there are several other places a namespace is created)
4. Use the unprefixed namespace everywhere
5. Use the prefixed namespace only on xpath queries (there are only three of these)

Example creation:

```java
this.defaultFitsNamespace = fitsXml.getRootElement().getNamespace();
        this.prefixedFitsNamespace = Namespace.getNamespace("fits", defaultFitsNamespace.getURI());
```

The reasoning here is that you normally want you operations to be using the same namespace definition as is on the root node. The only exception to this is in your xqueries, where you're explicitly binding a prefix in the query.

**Review by awoods — APPROVED** [review]
Looks good. Tests pass.

**Pull request merged**